backend/fms_core/services/report.py

```diff
@@ -0,0 +1,295 @@
+from typing import List
+import datetime
+from collections import defaultdict
+
+from typing import TypedDict, Union
+
+import pandas as pd
+from django.apps import apps
+from django.db.models import F, Count,  Sum, Max, Min, TextChoices, functions, QuerySet
+
+from io import BytesIO
+from openpyxl import Workbook
+from openpyxl.styles import PatternFill
+
+from fms_report.models import Report, MetricField
+from fms_report.models._constants import AggregationType, FieldDataType
+
+class TimeWindow(TextChoices):
+    MONTHLY = "month", "Monthly"
+    WEEKLY = "week", "Weekly"
+    DAILY = "day", "Daily"
+
+REPORT_HEADER_COLOR = "34a8eb"
+
+class EnhancedName(TypedDict):
+    name: str
+    display_name: str
+
+class ReportInfo(TypedDict):
+    report: EnhancedName
+    groups: list[EnhancedName]
+    time_windows: list[str]
+
+class ReportHeader(TypedDict):
+    name: str
+    display_name: str
+    field_order: int
+    aggregation: str
+    data_type: str
+
+class TimeWindowData(TypedDict):
+    time_window: datetime.date
+    time_window_label: str
+    time_window_start: datetime.date
+    time_window_end: datetime.date
+    time_window_data: list[dict[str, Union[str, int]]]
+    
+class ReportData(TypedDict):
+    report: EnhancedName
+    start_date: str
+    end_date: str
+    time_window: str
+    grouped_by: list[str]
+    headers: list[ReportHeader]
+    data: list[TimeWindowData]
+    
+
+def get_date_range_with_window(start_date: str, end_date: str, time_window: TimeWindow) -> tuple[list[str], list[str]]:
+    """
+    Provides two lists of dates. One list with each date between two given dates and one list with the matching time window date identifier for each of these dates.
+
+    Args:
+        `start_date`: start date for the list of dates.
+        `end_date`: end date for the list of dates.
+        `time_window`: size of the report page in time.
+  
+    Returns:
+        Tuple with one ordered date list and one matching time window list.
+    """
+    time_series = pd.date_range(start=start_date, end=end_date).to_series()
+    match time_window:
+        case TimeWindow.MONTHLY: 
+            window_time_series = time_series - pd.to_timedelta(time_series.dt.day - 1, unit="D")
+        case TimeWindow.WEEKLY:
+            window_time_series = time_series - pd.to_timedelta(time_series.dt.dayofweek, unit="D")
+        case TimeWindow.DAILY:
+            window_time_series = time_series
+
+    date_range = [datetime.date().isoformat() for datetime in pd.to_datetime(time_series, unit='D')]
+    time_window_range = [datetime.date().isoformat() for datetime in pd.to_datetime(window_time_series, unit='D')]
+    return date_range, time_window_range
```

Comment:
I've having brainfarts trying to understand this function.

Reply:
this was just to extract that logic from the main function. It returns a list of dates and a list of the matching time window identifier date. I'm glad we are working remote ...